File: ios/VidyoReactNative/VidyoConnectorManager.m
#import <UIKit/UIKit.h>
#import <Lmi/VidyoClient/VidyoConnector_Objc.h>
#import <React/RCTViewManager.h>
#import "RNNotification.h"

@interface VidyoConnectorManager : RCTViewManager<VCConnectorIConnect>
  @property (nonatomic, strong) UIView          *videoView;
  @property (nonatomic, strong) VCConnector     *vidyoConnector;
  @property (nonatomic, strong) RNNotification  *emitter;
@end

@implementation VidyoConnectorManager
RCT_EXPORT_MODULE();

- (UIView *)view
{
  _emitter    = [RNNotification allocWithZone: nil];
  _videoView  = [[UIView alloc] init];
  
  return _videoView;
}

RCT_EXPORT_METHOD(create:(int)remoteParticipants
           logFileFilter:(NSString *)logFileFilter
             logFileName:(NSString *)logFileName
                userData:(int)userData
                resolver:(RCTPromiseResolveBlock)resolve
                rejecter:(RCTPromiseRejectBlock)reject)
{
  dispatch_async(dispatch_get_main_queue(), ^{
    @try {
      bool initialized    = [VCConnectorPkg vcInitialize];
      self.vidyoConnector = [[VCConnector alloc] init:(void*)&_videoView
                                            ViewStyle:VCConnectorViewStyleDefault
                                   RemoteParticipants:remoteParticipants
                                        LogFileFilter:[logFileFilter UTF8String]
                                          LogFileName:[logFileName UTF8String]
                                             UserData:userData];
      if (initialized) {
        resolve(@"true");
      } else {
        resolve(@"false");
      }
    }
    @catch (NSError *error) {
      reject(@"Initialization_Error", @"Connector: creating error", error);
    }
  });
}

RCT_EXPORT_METHOD(destroy)
{
  dispatch_async(dispatch_get_main_queue(), ^{
    [_vidyoConnector disable];
    _videoView      = nil;
    _vidyoConnector = nil;
    [VCConnectorPkg uninitialize];
  });
}

RCT_EXPORT_METHOD(showViewCustom:(float)x
                               Y:(float)y
                           Width:(float)width
                          Height:(float)height)
{
  dispatch_async(dispatch_get_main_queue(), ^{
    int screenWidth  = [[UIScreen mainScreen] bounds].size.width;
    int screenHeight = [[UIScreen mainScreen] bounds].size.height;
    [_vidyoConnector showViewAt:&_videoView
                              X:     x * screenWidth
                              Y:     y * screenHeight
                          Width: width * screenWidth
                         Height:height * screenHeight];
  });
}

RCT_EXPORT_METHOD(connect:(NSString *)host
                    Token:(NSString *)token
              DisplayName:(NSString *)displayName
               ResourceId:(NSString *)resourceId)
{
  [_vidyoConnector connect:[host        cStringUsingEncoding:NSASCIIStringEncoding]
                     Token:[token       cStringUsingEncoding:NSASCIIStringEncoding]
               DisplayName:[displayName cStringUsingEncoding:NSASCIIStringEncoding]
                ResourceId:[resourceId  cStringUsingEncoding:NSASCIIStringEncoding]
         ConnectorIConnect:self];
}

RCT_EXPORT_METHOD(disconnect)
{
  [_vidyoConnector disconnect];
}

RCT_EXPORT_METHOD(setCameraPrivacy:(BOOL)cameraPrivacy)
{
  [_vidyoConnector setCameraPrivacy:cameraPrivacy];
}

RCT_EXPORT_METHOD(setMicrophonePrivacy:(BOOL)microphonePrivacy)
{
  [_vidyoConnector setMicrophonePrivacy:microphonePrivacy];
}

RCT_EXPORT_METHOD(getVersion:(RCTResponseSenderBlock)callback)
{
  NSString *clientVersion = [_vidyoConnector getVersion];
  callback(@[clientVersion]);
}

RCT_EXPORT_METHOD(setBackgroundMode)
{
  if (_vidyoConnector) {
    [_vidyoConnector selectLocalCamera:nil];
    [_vidyoConnector selectLocalMicrophone:nil];
    [_vidyoConnector selectLocalSpeaker:nil];
    [_vidyoConnector setMode:VCConnectorModeBackground];
  }
}

RCT_EXPORT_METHOD(setForegroundMode)
{
  if (_vidyoConnector) {
    [_vidyoConnector setMode:VCConnectorModeForeground];
    [_vidyoConnector selectDefaultCamera];
    [_vidyoConnector selectDefaultMicrophone];
    [_vidyoConnector selectDefaultSpeaker];
  }
}


-(void) onSuccess {
  [_emitter sendNotificationToReactNative:@"Connect:onSuccess" body:@"Successfully connected"];
}

-(void) onFailure:(VCConnectorFailReason)reason {
  [_emitter sendNotificationToReactNative:@"Connect:onFailure" body:@"Connection attempt failed"];
}

-(void) onDisconnected:(VCConnectorDisconnectReason)reason {
  if (reason == VCConnectorDisconnectReasonDisconnected) {
    [_emitter sendNotificationToReactNative:@"Connect:onDisconnected" body:@"Succesfully disconnected"];
  } else {
    [_emitter sendNotificationToReactNative:@"Connect:onDisconnected" body:@"Unexpected disconnection"];
  }
}

@end

File: android/app/src/main/java/com/vidyoreactnative/VidyoConnector/VidyoConnectorManager.java
package com.vidyoreactnative.VidyoConnector;

import android.Manifest;
import android.app.Activity;
import android.os.Build;
import android.support.v4.app.ActivityCompat;
import android.util.DisplayMetrics;
import android.widget.FrameLayout;

import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.vidyo.VidyoClient.Connector.ConnectorPkg;
import com.vidyo.VidyoClient.Connector.Connector;
import com.vidyoreactnative.NativeUIComponents.FrameLayoutManager;

public class VidyoConnectorManager extends ReactContextBaseJavaModule implements
        Connector.IConnect {

    public static final String REACT_CLASS = "VidyoConnectorManager";

    private static final String[] mPermissions = {
            Manifest.permission.CAMERA,
            Manifest.permission.RECORD_AUDIO,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
    };

    private ReactApplicationContext mReactContext;
    private Connector mVidyoConnector;
    private FrameLayout mVideoFrame;
    private FrameLayoutManager mLayoutManager;
    private Activity mActivity;
    private DisplayMetrics displayMetrics;

    public VidyoConnectorManager(ReactApplicationContext reactContext, FrameLayoutManager layout){
        super(reactContext);
        mReactContext  = reactContext;
        mLayoutManager = layout;

        displayMetrics = reactContext.getResources().getDisplayMetrics();
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @ReactMethod
    public void create(int remoteParticipants, String logFileFilter, String logFileName, Integer userData, Promise promise){
        try {
            mActivity   = getCurrentActivity();
            mVideoFrame = mLayoutManager.getFrameLayout();
            ConnectorPkg.setApplicationUIContext(mActivity);
            boolean initialized = ConnectorPkg.initialize();

            if (initialized && mVidyoConnector == null) {
                if (Build.VERSION.SDK_INT > 22) {
                    ActivityCompat.requestPermissions(mActivity, mPermissions, 1);
                }
                mVidyoConnector = new Connector(mVideoFrame,
                                                Connector.ConnectorViewStyle.VIDYO_CONNECTORVIEWSTYLE_Default,
                                                remoteParticipants,
                                                logFileFilter,
                                                logFileName,
                                                userData.longValue());
            }
            promise.resolve(initialized);
        } catch(Exception e) {
            promise.reject(e);
        }
    }

    @ReactMethod
    public void destroy() {
        mVidyoConnector.disable();
        mVidyoConnector = null;
        ConnectorPkg.setApplicationUIContext(null);
        ConnectorPkg.uninitialize();
    }

    @ReactMethod
    public void showViewDefault(){
        mVidyoConnector.showViewAt(mVideoFrame, 0, 0, displayMetrics.widthPixels, displayMetrics.heightPixels);
    }

    @ReactMethod
    public void showViewCustom(Float x, Float y, Float width, Float height){
        if (mVidyoConnector != null) {
            int marginLeft = (int) (displayMetrics.widthPixels * x);
            int marginTop = (int) (displayMetrics.heightPixels * y);
            int layoutWidth = (int) (displayMetrics.widthPixels * width);
            int layoutHeight = (int) ((displayMetrics.heightPixels * height) * 0.97);   // - status bar height ~3%

            mVidyoConnector.showViewAt(mVideoFrame, marginLeft, marginTop, layoutWidth, layoutHeight);
        }
    }

    @ReactMethod
    public void connect(String host, String token, String displayName, String resourceId){
        mVidyoConnector.connect(host, token, displayName, resourceId, this);
    }

    @ReactMethod
    public void disconnect() {
        mVidyoConnector.disconnect();
    }

    @ReactMethod
    public void setCameraPrivacy(Boolean mCameraPrivacy) {
        mVidyoConnector.setCameraPrivacy(mCameraPrivacy);
    }

    @ReactMethod
    public void setMicrophonePrivacy(Boolean mMicrophonePrivacy) {
        mVidyoConnector.setMicrophonePrivacy(mMicrophonePrivacy);
    }

    @ReactMethod
    public void getVersion(Callback version) {
        if (mVidyoConnector != null) {
            final String clientVersion = mVidyoConnector.getVersion();
            version.invoke(clientVersion);
        }
    }

    @ReactMethod
    public void setBackgroundMode() {
        //
    }

    @ReactMethod
    public void setForegroundMode() {
        //
    }

    // Handle successful connection.
    @Override
    public void onSuccess() {
        mReactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                     .emit("Connect:onSuccess", null);
    }

    // Handle attempted connection failure.
    @Override
    public void onFailure(Connector.ConnectorFailReason reason) {
        mReactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                     .emit("Connect:onFailure", "Connection attempt failed");
    }

    // Handle an existing session being disconnected.
    @Override
    public void onDisconnected(Connector.ConnectorDisconnectReason reason) {
        if (reason == Connector.ConnectorDisconnectReason.VIDYO_CONNECTORDISCONNECTREASON_Disconnected) {
            mReactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                         .emit("Connect:onDisconnected", "Succesfully disconnected");
        } else {
            mReactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                         .emit("Connect:onDisconnected", "Unexpected disconnection");
        }
    }

}
